hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/security/RootCARotationManager.java

```diff
@@ -255,11 +489,210 @@ public Duration timeBefore2ExpiryGracePeriod(X509Certificate certificate) {
+          scm.shutDown(message);
+        }
+
+        handler.setSubCACertId(newCertSerialId);
+
+        releaseLockOnTimeoutTask = executorService.schedule(() -> {
+          // If no rotation commit request received after rotation prepare
+          LOG.warn("Failed to have enough rotation acks from SCM. This " +
+              " time root rotation {} is failed. Release the lock.",
+              rootCACertId);
+          releaseLock();
+        }, ackTimeout.toMillis(), TimeUnit.MILLISECONDS);
+
+      } catch (Throwable e) {
+        LOG.error("Unexpected error happen", e);
+        scm.shutDown("Unexpected error happen, " + e.getMessage());
+      }
+    }
+  }
+
+  /**
+   *  Task to wait the all acks of prepare request.
+   */
+  public class WaitSubCARotationPrepareAckTask implements Runnable {
+    private String rootCACertId;
+
+    public WaitSubCARotationPrepareAckTask(
+        X509CertificateHolder rootCertHolder) {
+      this.rootCACertId = rootCertHolder.getSerialNumber().toString();
+    }
+
+    @Override
+    public void run() {
+      if (!isRunning()) {
+        LOG.info("SCM is not leader anymore. Delete the in-progress " +
+            "root CA directory");
+        cleanupAndStop("SCM is not leader anymore");
+        return;
+      }
+      synchronized (RootCARotationManager.class) {
+        int numFromHADetails =
+            scm.getSCMHANodeDetails().getPeerNodeDetails().size() + 1;
+        int numFromRatisServer = scm.getScmHAManager().getRatisServer()
+            .getDivision().getRaftConf().getCurrentPeers().size();
+        LOG.info("numFromHADetails {}, numFromRatisServer {}",
+            numFromHADetails, numFromRatisServer);
+        if (handler.rotationPrepareAcks() == numFromRatisServer) {
+          // all acks are received.
+          try {
+            waitAckTimeoutTask.cancel(false);
+            handler.rotationCommit(rootCACertId);
+            handler.rotationCommitted(rootCACertId);
+
+            metrics.incrSuccessRotationNum();
+            long timeTaken = System.nanoTime() - processStartTime.get();
+            metrics.setSuccessTimeInNs(timeTaken);
+            processStartTime.set(null);
+
+            // reset state
+            handler.resetRotationPrepareAcks();
+            String msg = "Root certificate " + rootCACertId +
+                " rotation is finished successfully after " + timeTaken + " ns";
+            cleanupAndStop(msg);
+          } catch (Throwable e) {
+            LOG.error("Execution error", e);
+            handler.resetRotationPrepareAcks();
+            cleanupAndStop("Execution error, " + e.getMessage());
+          } finally {
+            // stop this task to re-execute again in next cycle
+            throw new RuntimeException("Exit the this " +
+                "WaitSubCARotationPrepareAckTask for root certificate " +
+                rootCACertId + " since the rotation is finished execution");
```

Comment:
Why are we throwing exception from `finally` block here?

Reply:
I guess we want to stop scheduling this task again...

I think we can take the reference of the Future once the task is submitted, and cancel via that one instead.
Nice catch @nandakumar131 we have talked about this with Sammi, I just missed this one from my notes as it seems. Similarly as we cancel the waitAckTimeoutTask.